script: Implement <input type=range> widget's layout

BudiArb opened the pull request:
Implement the layout of `<input type="range">`. Support automatically calculating the position of thumb based on the current input value, and the size of range-track and range-thumb. Added a document.has_script_or_layout_blocker function to detect whether it is possible to run box_area_query during bind_to_tree, add delay_task if there are blocker.

Here are some of the fixes made in this PR:
1. Fixed the structure for input type range's pseudo elements.
2. Fixed the positioning of input type range's thumb based on current value, width, and direction.
3. Allow input type range to stretch vertically in a bigger container.

Original PR: #41024
Stylo PR: https://github.com/servo/stylo/pull/310

Testing: Some improvements in WPT tests, with a few regressions. This change includes a Servo-specific appearance test to detect unexpected changes to the look and feel of range widgets.
Fixes: #22728

rayguo17 committed:
implement  <input type=range> widget's layout, automatically calculate the position of thumb based on value.

Co-authored-by: Tin Tun Aung <tin.tun.aung1@huawei.com>
Signed-off-by: Budiman Arbenta <arbenta6@gmail.com>

shubhamg13 reviewed (commented):
It needs to be rebased as it has conflicts. Please rebase.

shubhamg13 commented on tests/wpt/meta/css/css-grid/alignment/grid-self-alignment-stretch-input-range.html.ini:
```diff
@@ -0,0 +1,2 @@
+[grid-self-alignment-stretch-input-range.html]
+  expected: FAIL
```
nit: Add extra line at EOF.

I think you updated the expectations manually. You can automate creation of `.ini` files by running it in specific folder by using [this](https://book.servo.org/contributing/testing.html#updating-web-platform-test-expectations)

mrobinson commented on components/script/dom/html/htmlinputelement.rs:
```diff
@@ -1337,6 +1346,96 @@ impl HTMLInputElement {
+    fn create_range_shadow_tree(&self, can_gc: CanGc) {
+        let document = self.owner_document();
+        let shadow_root = self.shadow_root(can_gc);
+        Node::replace_all(None, shadow_root.upcast::<Node>(), can_gc); // Clear all the child inside.
```
```suggestion
        Node::replace_all(None, shadow_root.upcast::<Node>(), can_gc);
```

mrobinson reviewed (commented):
Initial comments, but looks pretty straightforward to me.

rayguo17 commented:
> In addition, this change makes it so that `<input type=range>` is not interactive at all. This is probably fine, but we need to implement a quick followup which adds both mouse and keyboard interactivity for the control.

I have a issue tracking the interactivity of `<input type=range>` https://github.com/servo/servo/issues/40759

xiaochengh commented on components/script/dom/html/htmlinputelement.rs:
```diff
@@ -316,14 +316,140 @@ impl ColorInputShadowTree {
+            percent,
+            thumb_rect_width * percent / 100.0
+        );
+        let progress_style = format!("width: {percent}%;");
```
```suggestion
        let progress_style = format!("width: {percent}% !important;");
```
Otherwise it can be overwritten by author style

xiaochengh reviewed (commented):
Could you also add some tests to make sure the appearance is correct?

Example: https://github.com/servo/servo/blob/main/tests/wpt/mozilla/tests/appearance/input-textual-text-align.html

stevennovaryo reviewed (commented):
Just a little bit of lookaround! Hope to looks at the details of CSS styling later. Note that this PR would be affected a lot by the https://github.com/servo/servo/pull/43325 `HTMLInputElement` refactoring. 

Also having a private Servo specific appearance WPTs would be helpful. Since otherwise we have no way to detect whether the appearance of this kind of inputs is correct per commits.

BudiArb committed:
Add test for appearance check and clear some nit

Signed-off-by: Budiman Arbenta <arbenta6@gmail.com>

BudiArb committed:
Remove the need to wait for layout blocker on range input update

Signed-off-by: Budiman Arbenta <arbenta6@gmail.com>

BudiArb commented on components/script/dom/html/htmlinputelement.rs:
```diff
@@ -316,14 +316,140 @@ impl ColorInputShadowTree {
+                }
+            ));
+            return;
+        }
```
I just found a way to update the position without needing the size of the thumb. I have removed these lines of code

mrobinson committed:
Small improvements

 - Update the `Cargo.toml` to reference the Stylo referencej
 - Update the style to match the rest of Servo's form controls more
   closely and react to hover and active.

Signed-off-by: Martin Robinson <mrobinson@igalia.com>
